### PR TITLE
feat: Add a way to see all permissions required to run a given Command

### LINKED ISSUE #511: Ability to call a given command's required permission
### Summary

It just returns what's in `@commands.has_permission()`.

### What is the feature request for?

nextcord.ext.commands

### The Problem

It'd be cool to have this in a help command that allows you to do `!help [command]`.

### The Ideal Solution

Probably something like `Command.required_permission`.

### The Current Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Skelmis**
## Summary

This adds the ability to view the permissions required to run a command. 

As an implementation detail, this method also ensures the immutability of the decorator, ensuring the permissions output stays in line with the required permissions despite runtime changes. 

It would have been nice to avoid setting the attribute each time the check runs, however, due to how the checks work internally it appears you cannot triple nest decorator functions to get that functionality. If you are aware of a way to do this, please comment with a worked nextcord example.

Closes #511 

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

## Testing

These tests show both the usage and the built-in immutability of this approach. Such that even after 'removing' the permissions, they are still set correctly on the next invocation.
```python
import os

from nextcord.ext import commands
from nextcord.ext.commands import Context

bot = commands.Bot(command_prefix="$")


@bot.command()
@commands.has_permissions(manage_guild=True)
async def ping(ctx: Context):
    await ctx.send(str(ctx.command.required_permissions))
    ctx.command.callback.required_permissions = {}
    await ctx.send(str(ctx.command.required_permissions))


@bot.command()
@commands.has_guild_permissions(manage_guild=True)
async def ping_2(ctx: Context):
    await ctx.send(str(ctx.command.required_guild_permissions))
    ctx.command.callback.required_guild_permissions = {}
    await ctx.send(str(ctx.command.required_guild_permissions))


@bot.command()
@commands.bot_has_permissions(manage_guild=True)
async def ping_3(ctx: Context):
    await ctx.send(str(ctx.command.required_bot_permissions))
    ctx.command.callback.required_bot_permissions = {}
    await ctx.send(str(ctx.command.required_bot_permissions))


@bot.command()
@commands.bot_has_guild_permissions(manage_guild=True)
async def ping_4(ctx: Context):
    await ctx.send(str(ctx.command.required_bot_guild_permissions))
    ctx.command.callback.required_bot_guild_permissions = {}
    await ctx.send(str(ctx.command.required_bot_guild_permissions))

bot.run(os.environ["TOKEN"])
```
image
image

### CODE REVIEW

File: nextcord/ext/commands/core.py [hunk 1]
```diff
@@ -423,6 +423,73 @@ def __init__(self, func: Union[
 
         self.checks.extend(self.parent.checks)  # type: ignore
 
+    # Issue #511
+    @property
+    def required_permissions(self) -> Dict[str, bool]:
+        """
```
Comment (ooliver1) [inline nextcord/ext/commands/core.py:429]: style comment: consistency and PEP257 - brief description on the same line as the opening `"""`

File: nextcord/ext/commands/core.py [hunk 2]
```diff
@@ -423,6 +423,73 @@ def __init__(self, func: Union[
...
+        Returns the permissions required to run this command.
+
+        Returns
+        -------
+        dict
```
Comment (ooliver1) [inline nextcord/ext/commands/core.py:434]:
```suggestion
        Dict[:class:`str`, :class:`bool`]
```
for specificity in docs

**Review by ooliver1 — CHANGES_REQUESTED** [review]

**Comment by ooliver1** [comment]
and ^ for all 4 of course

**Review by ooliver1 — APPROVED** [review]

**Review by ooliver1 — APPROVED** [review]

**Comment by ooliver1** [comment]
image
there seems to be some conflicts, im getting ` ! [rejected]          refs/pull/618/head -> fix/511  (non-fast-forward)` and unable to edit in github so I assume you did not enable `allow maintainers to make modifications`, either enable that or resolve conflicts

**Comment by Skelmis** [comment]
That darn black pr haha, however you have access now. I will resolve it regardless

**Comment by Skelmis** [comment]
Not sure what you want me to do about pyright, outside of just type ignoring them

**Comment by Skelmis** [comment]
TODO: Also add this to application checks

**Comment by ooliver1** [comment]
> Not sure what you want me to do about pyright, outside of just type ignoring them

you could just conditionally set `if ctx.command is not None`

**Comment by Soheab** [comment]
Why was this closed?

**Pull request closed without merging**